### ISSUE TITLE
add rdflib-endpoint support to ODK?

### ISSUE THREAD

**Issue by matentzn** [INCATools/ontology-development-kit#855]
@gouttegd  shared with us some of his experiences with rdflib-endpoint in his _amazing_ OBO Academy tutorial on Protege 5.6.1.

Without voicing your opinion just yet @gouttegd, can you share your recipe here for how to use this for query development use cases etc? Is it hard to install? How to deploy?

**Comment by gouttegd** [comment]
Not sure adding `rdflib-endpoint` to the ODK would be useful. `rdflib-endpoint` works by creating a web server on the localhost, that you can connect to using your web browser.

Running `rdflib-endpoint` from within an ODK container would confine the web server to that container, where it would not be accessible from the host’s web browser – unless you configure Docker to allow networking between the container and the host, which is doable but adds some noticeable complexity.

`rdflib-endpoint` is easily installable via PyPI, so I am not too keen on increasing the ODK complexity just to save users a mere `python -m pip install rdflib-endpoint`…

**Comment by gouttegd** [comment]
As for my “recipe”:

1. Install `rdflib-endpoint`:

```sh
$ python -m pip install rdflib-endpoint
```

2. Start the server with the file you want to query against. Note that `rdflib-endpoint` does not support XML catalogs AFAIK, so it’s best not to use it against a ODK-style `-edit` file, otherwise it will get the imports from their actual remote locations and _not_ from the local paths configured in the catalog.

```sh
$ rdflib-endpoint serve cl-base.owl
INFO:     📥️ Loaded triples from ../../cl-base.owl, for a total of 71821
INFO:     Started server process [25]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
```

3. Point your web browser to the indicated address (`http://0.0.0.0:8000`) and start querying.

4. When you’re done, terminate the server by sending a `^C` as indicated.

**Comment by gouttegd** [comment]
Coming back to this issue.

We _could_ “add `rdflib-endpoint` support to the ODK”. What this would entail:

1) Add the `rdflib-endpoint` and `uvicorn` Python packages to the ODK image (`uvicorn` is a runtime dependency of `rdflib-endpoint`, but for some reason it is not explicitly listed as such, so its installation must be explicitly required) – that’s completely trivial.

2) Tweak the launcher script (or my `odkrun`) to start the ODK container while exposing port 8000 inside the container to a port in localhost (e.g. `docker run --network host --publish 127.0.0.1:8000:8000/tcp`), and inside the image spawn the command `rdflib-endpoint serve file.owl`.

While the second point is certainly doable, I am reluctant to add yet more complexity to that poor `run.sh` script – and before I add network-tweaking code to my `odkrun`, I’d like to know this is going to be useful and/or fulfill someone’s needs. But the complete lack of engagement on this ticket does not make me think this is a feature that is awaited by many people, if any.

So in short, I believe the feature is not worth the added complexity to the runner (both the `run.sh` script and my `odkrun` program). Therefore I am inclined to close here and to optionally transform this issue into a documentation issue on the OBOOK, to teach people how they can use `rdflib-endpoint` on their own machine without needing the ODK.

**Comment by matentzn** [comment]
I agree, there probably should be more pressure from users before proceeding. I am fine with closing it.
